content/style/go.md

```diff
@@ -1,3 +1,1659 @@
+## Guidelines
+
+### Copy Slices and Maps at Boundaries
+
+Slices and maps contain pointers to the underlying data so be wary of scenarios when they need to be copied.
+
+#### Receiving Slices and Maps
+
+Keep in mind that users can modify a map or slice you received as an argument if you store a reference to it.
+
+<table>
+<thead><tr><th>Bad</th> <th>Good</th></tr></thead>
+<tbody>
+<tr>
+<td>
+
+```go
+func (d *Driver) SetTrips(trips []Trip) {
+  d.trips = trips
+}
+
+trips := ...
+d1.SetTrips(trips)
+
+// Did you mean to modify d1.trips?
+trips[0] = ...
+```
+
+</td>
+<td>
+
+```go
+func (d *Driver) SetTrips(trips []Trip) {
+  d.trips = make([]Trip, len(trips))
+  copy(d.trips, trips)
+}
+
+trips := ...
+d1.SetTrips(trips)
+
+// We can now modify trips[0] without affecting d1.trips.
+trips[0] = ...
+```
+
+</td>
+</tr>
+
+</tbody>
+</table>
+
+#### Returning Slices and Maps
+
+Similarly, be wary of user modifications to maps or slices exposing internal state.
+
+<table>
+<thead><tr><th>Bad</th><th>Good</th></tr></thead>
+<tbody>
+<tr><td>
+
+```go
+type Stats struct {
+  mu sync.Mutex
+  counters map[string]int
+}
+
+// Snapshot returns the current stats.
+func (s *Stats) Snapshot() map[string]int {
+  s.mu.Lock()
+  defer s.mu.Unlock()
+
+  return s.counters
+}
+
+// snapshot is no longer protected by the mutex, so any
+// access to the snapshot is subject to data races.
+snapshot := stats.Snapshot()
+```
+
+</td><td>
+
+```go
+type Stats struct {
+  mu sync.Mutex
+  counters map[string]int
+}
+
+func (s *Stats) Snapshot() map[string]int {
+  s.mu.Lock()
+  defer s.mu.Unlock()
+
+  result := make(map[string]int, len(s.counters))
+  for k, v := range s.counters {
+    result[k] = v
+  }
+  return result
+}
+
+// Snapshot is now a copy.
+snapshot := stats.Snapshot()
+```
+
+</td></tr>
+</tbody></table>
+
+### Start Enums at One
+
+The standard way of introducing enumerations in Go is to declare a custom type and a `const` group with `iota`.
+Since variables have a 0 default value, you should usually start your enums on a non-zero value.
+
+<table>
+<thead><tr><th>Bad</th><th>Good</th></tr></thead>
+<tbody>
+<tr><td>
+
+```go
+type Operation int
+
+const (
+  Add Operation = iota
+  Subtract
+  Multiply
+)
+
+// Add=0, Subtract=1, Multiply=2
+```
+
+</td><td>
+
+```go
+type Operation int
+
+const (
+  Add Operation = iota + 1
+  Subtract
+  Multiply
+)
+
+// Add=1, Subtract=2, Multiply=3
+```
+
+</td></tr>
+</tbody></table>
+
+There are cases where using the zero value makes sense, for example when the zero value case is the desirable default behavior.
+
+```go
+type LogOutput int
+
+const (
+  LogToStdout LogOutput = iota
+  LogToFile
+  LogToRemote
+)
+
+// LogToStdout=0, LogToFile=1, LogToRemote=2
+```
+
+### Error Types
+
+There are various options for declaring errors:
+
+* [`errors.New`] for errors with simple static strings
```

Comment:
It's because they are linked. You can do the following in Markdown:

```markdown
tata [titi] tutu

     [titi]: https://google.com
```

It's neat for big documents like this one to keep the sentences readable and the document neat. WDYT?

Reply:
Ah, I didn't know that about markdown, that's super neat :)

I can see that works in `go.md` when I view it as markdown, but not in the page I get when I `mkdocs serve`. Does it work in the compiled version? (Does that differ from what I get with just serving it?)

Reply:
Aha, so currently it seems like Mkdocs supports reference-style links (this is what it's called) but not implicit ones, where the key and the link's content are the same. So for it to work we would have to write:

```markdown
tata [titi] [titi_link] tutu

     [titi_link]: https://google.com
```

Which kind of sucks. I'm looking if there are some extensions that would allow this, but otherwise I'm not sure what to do since I don't like reference-style links that are verbose, and I also don't like unmaintainable inline links for such huge documents.

Reply:
So it seems they are supported but using backticks in them breaks the links. Should be an easy fix then, I'll just not use inline code blocks within implicit reference-style links! (that was a mouthful!)